TypeCobol/Compiler/Preprocessor/ReplaceTokensLinesIterator.cs

```diff
@@ -517,33 +504,80 @@ internal static Token GenerateReplacementToken(Token originalToken, string repla
         /// <summary>
         /// Rescan the TokenType of a set of replaced Tokens.
         /// </summary>
+        /// <param name="creator">A function that create a Replaced Token from a replacement</param>
         /// <param name="firstOriginalToken">The first original token to be replaced</param>
         /// <param name="replacedTokens">The array of replacement tokens.</param>
-        private static void RescanReplacedTokenTypes(Token firstOriginalToken, params Token[] replacedTokens)
+        /// <returns>An array of new calculated replaced tokens</returns>
+        private static T[] RescanReplacedTokenTypes<T>(Func<Token, T> creator, Token firstOriginalToken, params Token[] replacementTokens)
+            where T : Token
         {
-            MultilineScanState scanState = firstOriginalToken.ScanStateSnapshot;
-            if (scanState != null && replacedTokens.Any(MultilineScanState.IsScanStateDependent))
+            MultilineScanState scanState = firstOriginalToken?.ScanStateSnapshot;
+            if (scanState != null && replacementTokens.Any(MultilineScanState.IsScanStateDependent))
             {
+                int i = 0;
+                int[] columns = new int[replacementTokens.Length + 1];                
                 StringBuilder sb = new StringBuilder();
-                foreach (var t in replacedTokens)
+                foreach (var t in replacementTokens)
                 {
-                    sb.Append(t.Text);
-                    sb.Append(' ');
+                    columns[i++] = sb.Length;
+                    sb.Append(t.Text);                    
                 }
+                columns[i] = sb.Length;
                 string tokenText = sb.ToString();
+                int startTokIdx = 0;
+                int endTokIdx = 0;                
+                List<T> newReplacedTokens = new List<T>(replacementTokens.Length);
                 TokensLine tempTokensLine = TokensLine.CreateVirtualLineForInsertedToken(0, tokenText);
                 tempTokensLine.InitializeScanState(scanState);
                 var tempScanner = new TypeCobol.Compiler.Scanner.Scanner(tokenText, 0, tokenText.Length - 1, tempTokensLine, new TypeCobolOptions(), true);
                 Token rescannedToken = null;
-                int i = 0;
+                List<Token> tokens = new List<Token>((replacementTokens.Length >> 1) + 1);
```

Comment:
Why setting this capacity ? Why using bit shifting operator ?

Reply:
I just wanted to optizme space here, tokens is used for gathering Tokens Groups, so in most cases grouped tokens will never exceed (ReplacementToken.Length / 2) + 1; Right shift operator is the same than dividing by 2.

Reply:
I know what the operator does but it is unusual to use it. I think it is clearer for the reader to see the divide by 2 operation explicitly. Or add a comment to explain this optimization.